src/main/java/duke/task/TaskList.java

```diff
@@ -0,0 +1,149 @@
+     * marked as done.
+     */
+    public String markDone(int taskNumber) {
+        String message = "";
+
+        try {
+            Task completedTask = tasks.get(taskNumber);
+            completedTask.markDone();
+            message = Ui.getDoneMessage(completedTask);
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("OOPS!!! This is not a valid task number.");
+        }
+        return message;
+    }
+
+    /**
+     * Prints the entire TaskList if the TaskList is not empty. If the TaskList is empty,
+     * a message that informs the user that they do not have any tasks is printed.
+     *
+     * @return A String of the list of tasks in the task list. If the user does not have
+     * tasks, the String returned will inform the user accordingly.
+     */
+    public String listTasks() {
+        StringBuilder message = new StringBuilder();
+
+        if (!tasks.isEmpty()) {
+            message.append("Here are the tasks in your list:");
+            for (int i = 0; i < tasks.size(); i++) {
+                message.append("\n")
+                        .append(i + 1)
+                        .append(". ")
+                        .append(tasks.get(i));
+            }
+            return message.toString();
+        } else {
+            return "You don't have any tasks in your list!";
+        }
+    }
+
+    /**
+     * Adds a specified Task into the TaskList.
+     *
+     * @param task The specified Task to add into the TaskList.
+     * @return A String containing the message to be shown to the user when a task
+     * is added into the task list.
+     */
+    public String addTask(Task task) {
+        tasks.add(task);
+        return Ui.getAddedMessage(task, tasks.size());
+    }
+
+    /**
+     * Deletes a specified Task from the TaskList.
+     *
+     * @param taskNumber The task number of the Task that is to be deleted.
+     * @return A String containing the message to be shown to the user when a task
+     * is removed from the task list.
+     */
+    public String deleteTask(int taskNumber) {
+        try {
+            Task removedTask = tasks.get(taskNumber);
+            tasks.remove(taskNumber);
+            return Ui.getDeleteMessage(removedTask, tasks.size());
+        } catch (IndexOutOfBoundsException e) {
+            return "OOPS!!! This is not a valid task number.";
+        }
+    }
+
+    /**
+     * Prints all the Tasks in the taskList that contain the specified keyword.
+     *
+     * @param keyword The keyword specified by the user to search for.
+     * @return A String of the list of tasks found using the keyword.
+     */
+    public String findTask(String keyword) {
+        boolean canFindMatch = false;
+        StringBuilder message = new StringBuilder();
+
+        for (int i = 0; i < tasks.size(); i++) {
+            String currentTask = tasks.get(i).toString();
+            if (currentTask.substring(7).matches("(?i)(.*)" + keyword + "(.*)")) {
```

Comment:
I think you can improve this line by assign this line: `currentTask.substring(7).matches("(?i)(.*)" + keyword + "(.*)")` into a descriptive boolean variable so it is easier tor read :)